core/arch/aarch64/emit_utils.c

```diff
@@ -446,8 +446,14 @@ indirect_linkstub_stub_pc(dcontext_t *dcontext, fragment_t *f, linkstub_t *l)
     cache_pc cti = EXIT_CTI_PC(f, l);
```

Comment:
Can you check if any test in the current suite hits all the new cases introduced by this PR? Also add details of the manual testing you did in the PR's description.

Reply:
I haven't checked the test suite yet, but will do.

The manual test I did was to run spec cpu2017 under DR and currently if is able to run all of the benchmarks that can be run without trace enabled on AARCH64.

Reply:
I had a brief look at the test suite, and I haven't found any places where the newly introduced code would be explicitly tested (thinking about how a fragment would be mangled into another one with `str/mov` removed). I haven't found any code that tested `mangle_indirect_branch_in_trace()` for x86 either. (but maybe I missed something, if any one knows where the test for x86 trace building is, I would love to know)

There is one test that tested whether a trace is built if some bbs are executed frequently enough, in client.null_instrument.

I am wondering what would be the best way to proceed, wrote tests in assembly that verify the fragment mangling?

Reply:
I couldn't find much testing for traces in x86 either. I think it'll be good to add unit tests though because I see that the implementation sometimes depends on how other mangling (like `mangle_cbr_stolen_reg`) looks like.

> I am wondering what would be the best way to proceed, wrote tests in assembly that verify the fragment mangling?

Yeah, we do have tests that create basic blocks in assembly. E.g. see `drreg-test.c` (we don't need a separate test client for this PR though).

@derekbruening any other ideas? I don't know how we can verify in unit tests if DR indeed created a trace or not. I can only think of the "Trace fragments generated" stats. Do you think it's worth the effort adding unit tests for traces (asking because I couldn't find much testing for x86).

Reply:
I had an offline discussion with Derek, and it seems that currently we rely mostly on traces being built in the existing tests. This is of course not ideal as it would allow detection of only those regressions that cause crashes. We do want comprehensive tests for trace building that verify corner cases as well, but that's a lot of more work. I've created #5057 for that. For this PR, I think we should document trace creation related stats (like count of traces created and others if they are readily available in DR stats) for SPEC and our current test suite.

Reply:
Ok, so when you say we need to document trace creation related stats, what exactly do you mean by that? Do we need to have a file somewhere to log all the stats of trace creation, or just put some data in this PR?

Reply:
> as in where shall we document it

In this PR's description and the final commit message.

> I guess that we could use macros like GLOBAL_STAT(num_trace) in a call to LOG in the code?

Can you check which stats are already output in log files? We can extract those numbers and put in the PR's description.

Reply:
I had a quick look at the log file generated by DR using `-loglevel 4 -logmask LOG_STATS` and there are some relevant stats such as
```
                 Fragments generated, bb and trace (thread):              1981
                   Basic block fragments generated (thread):              1946
                         Trace fragments generated (thread):                35
               Trace building reset: no trace head (thread):                35
               Number of bbs in all emitted traces (thread):               123
                  Maximum number of bbs in a trace (thread):                11
        Trace wannabes prevented from being traces (thread):              2391
                       Shadowed trace head deleted (thread):                35
                             Trace heads re-marked (thread):                 8
                        Future fragments generated (thread):              2558
                        Shared fragments generated (thread):              1852
                              Shared bbs generated (thread):              1817
                           Shared traces generated (thread):                35

```
And here is the log file if you want to take a look.
[log.0.345299.txt](https://github.com/DynamoRIO/dynamorio/files/7039370/log.0.345299.txt)

Reply:
This looks useful. I suppose the stats in your above comment were for some small test app in the DR suite (as the counts seem low)? It'd be good to document these counts for SPEC as well.

Reply:
It is run on a small program, although not on the test suite. Here is some log information on a test suite of SPEC CPU2017
```
                 Fragments generated, bb and trace (thread):             61212
                   Basic block fragments generated (thread):             52009
                         Trace fragments generated (thread):              9203
               Trace building reset: no trace head (thread):              9203
               Number of bbs in all emitted traces (thread):             22522
                  Maximum number of bbs in a trace (thread):                22
        Trace wannabes prevented from being traces (thread):            100694
                       Shadowed trace head deleted (thread):              9203
                             Trace heads re-marked (thread):               118
                        Future fragments generated (thread):             68257
                        Shared fragments generated (thread):             38679
                              Shared bbs generated (thread):             29476
                           Shared traces generated (thread):              9203

```
And this is the stats on the test common.broadfun
```
              Fragments generated, bb and trace (thread):              1902
                   Basic block fragments generated (thread):              1856
                         Trace fragments generated (thread):                46
               Trace building reset: no trace head (thread):                46
               Number of bbs in all emitted traces (thread):               136
                  Maximum number of bbs in a trace (thread):                11
        Trace wannabes prevented from being traces (thread):              2622
                       Shadowed trace head deleted (thread):                46
                             Trace heads re-marked (thread):                 8
                        Future fragments generated (thread):              2422
                        Shared fragments generated (thread):              1761
                              Shared bbs generated (thread):              1715
                           Shared traces generated (thread):                46

```

Reply:
With the other tests seems fine.